### ISSUE TITLE
Reference counting leak in JProxy

### ISSUE THREAD

**Issue by Thrameos** [jpype-project/jpype#472]
JProxy needs to properly implement the GC container reference counting pattern.   Currently because an instance that calls `@JImplements` will have a `__javaproxy__` that points back to the instance object, they form a circular reference.

**Comment by Thrameos** [comment]
Example showing the issue.

``` python
import time
import gc
import sys
import jpype

jpype.startJVM()
from jpype import java, JImplements, JOverride, JObject

@JImplements(java.lang.Runnable)
class MyRunnable(object):
   @JOverride
   def run(self):
      pass

# Garbage collect anything left from loading
gc.collect()
gc.set_debug(gc.DEBUG_SAVEALL)

# Create some proxies
runnable = MyRunnable()
jrunnable = JObject(runnable, java.lang.Runnable)
run_id = id(runnable)
proxy_id = id(runnable.__javaproxy__)
jrun_id = id(jrunnable)

del runnable

# deleting the runnable will free if from Java
if True:
   del jrunnable

    # Java will hold a copy until we garbage collect
    java.lang.System.gc()
    time.sleep(2)
else:
    # Shutting down should also remove all references
    jpype.shutdownJVM()

# Call Python GC
gc.collect()
ids = set([id(item) for item in gc.garbage])

# Check if we got it
if run_id in ids:
  print("Collected runnable")
else:
  print("Fail to collect runnable")

if proxy_id in ids:
  print("Collected proxy")
else:
  print("Fail to collect proxy")

```

**Comment by Thrameos** [comment]
Fixed in #470
